### PR TITLE
Allocate Vec capacity based on size hints during decoding

### PR DESCRIPTION
**Pull request by jordy25519**
Discovered this issue while profiling another crate

### CODE REVIEW

**Review by vkgnosis — APPROVED** [review]

**Review by vkgnosis — CHANGES_REQUESTED** [review]
This fails CI. I guess there is a case where the code without with_capacity never end ups really allocating that much.

**Comment by jordy25519** [comment]
> This fails CI. I guess there is a case where the code without with_capacity never end ups really allocating that much.

Ok I'm glad this happened since some malicious payloads could cause large memory allocation in my previous commit. Iterator should have the same effect to avoid reallocs and not allocate upfront

**Comment by vkgnosis** [comment]
> Iterator should have the same effect to avoid reallocs and not allocate upfront

I don't think that's true. The current version of the PR works the same as it did before the PR but using `collect` instead of manually pushing. Either way the Vector reallocate as elements get added because the final size isn't known upfront.

File: ethabi/src/filter.rs [hunk 1]
```diff
@@ -82,12 +83,6 @@ impl<T> Topic<T> {
 	}
 }
 
-impl<T> Default for Topic<T> {
-	fn default() -> Self {
-		Topic::Any
-	}
-}
```
Comment (vkgnosis) [inline ethabi/src/filter.rs:89]: Why this change? I think you are changing behavior because the previous version still implemented Default when T wasn't Default while the new one doesn't. Either way, it's unrelated to the rest of the  PR.

Comment (jordy25519) [reply ethabi/src/filter.rs:89]: I only changed it due to the CI failure, happy to keep or revert it either way

**Review by vkgnosis — APPROVED** [review]
